_posts/2019-01-18-monorepos-aws-codebuild.markdown

```diff
@@ -0,0 +1,183 @@
+
+Our configuration to use jest in multi-project mode looks like this -
+
+```javascript
+module.exports = {
+    globals: {
+        "ts-jest": {
+            diagnostics: false,
+            isolatedModules: true
+        }
+    },
+    // Each path in the array below uses the jest configuration
+    // at that path for that particular 'project'.
+    projects: [
+        "<rootDir>/packages/package_one",
+        "<rootDir>/packages/package_two",
+        "<rootDir>/packages/package_three",
+        "<rootDir>/packages/package_four",
+    ],
+    testEnvironment: "node",
+    transform: {
+        "^.+\\.tsx?$": "ts-jest",
+    },
+    testPathIgnorePatterns: ["/lib/", "/node_modules/"],
+    moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
+    coverageDirectory: "./.cover",
+    collectCoverageFrom: ["**/src/**/*.{ts}", "!**/node_modules/**"],
+    coverageThreshold: {
+        "global": {
+            "branches": 80,
+            "functions": 80,
+            "lines": 80,
+            "statements": 80
+        }
+    }
+};
+```
+
+With this, we can specify executing a particular group of projects with the following command -
+
+```bash
+./node_modules/.bin/jest --ci "/tests/.*\\.(test|spec|integration)?\\.(ts|tsx)$" --projects ./packages/package_one ./packages/package_two
+```
+
+If we combine this with lerna changed, we can now determine which packages have changed since we made any commits to master, and test only those packages. In our case - the following command can be issued.
+
+```bash
+./node_modules/.bin/jest --ci "/tests/.*\\.(test|spec|integration)?\\.(ts|tsx)$" --projects $(./node_modules/.bin/lerna changed --all -p | grep -Eo -e packages/.+)
+```
+
+This allows us to target execute tests against only packages which have changed. This does require you to perform a full checkout of the repository in AWS Codebuild, as opposed to the default behaviour which is to perform a shallow clone. We don't currently have this activated as we had been waiting for Jest 24 to release in order to resolve some other unrelated issues. With Jest 24 having been recently launch, I look forward to turning this on!
+
+I hope this gives everyone a good look at the steps that need to be taken to keep the build process of a monorepo running reasonably efficiently. In this case, it has included;
+
+- Installing specialised tooling to manage a monorepo, in our case it was lerna.
+- Taking advantage of dependency caching in AWS CodeBuild.
+- Utilising npm ci to resolve and install dependencies.
+- Running jest in multi-project mode together with lerna changed.
+
+Hopefully this has been helpful at anyone who is looking at taking the Monorepo approach. 
```

Comment:
maybe add your contact details or mech rocks, if there are anyone experiencing similar build issues and needs help ;-)